### ISSUE TITLE
关于ENet-Label-Torch训练问题

### ISSUE THREAD

**Issue by ygqhfut** [cardwing/Codes-for-Lane-Detection#191]
作者您好，我只有一个1080Ti的GPU，在训练的时候，需要怎么修改test.sh脚本里的参数，我尝试着改了下，出现了这样的错误：
/home/hfut/torch/install/bin/luajit: /home/hfut/torch/install/share/lua/5.1/nn/Module.lua:286: bad argument #1 to 'set' (expecting number or Tensor or Storage)
stack traceback:
	[C]: in function 'set'
	/home/hfut/torch/install/share/lua/5.1/nn/Module.lua:286: in function 'flatten'
	/home/hfut/torch/install/share/lua/5.1/nn/Module.lua:326: in function 'getParameters'
	./train.lua:36: in function '__init'
	/home/hfut/torch/install/share/lua/5.1/torch/init.lua:91: in function </home/hfut/torch/install/share/lua/5.1/torch/init.lua:87>
	[C]: in function 'Trainer'
	main.lua:36: in main chunk
	[C]: in function 'dofile'
	...hfut/torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:150: in main chunk
	[C]: at 0x00405d50
从网上没有找到相应的解决方案，最后期待作者您的回复。

**Comment by cardwing** [comment]
你的batch size是多少？我记得batch size是1的时候会有问题，如果你的batch size刚好设为1的话就增大一下batch size就好了。

**Comment by ygqhfut** [comment]
> 你的batch size是多少？我记得batch size是1的时候会有问题，如果你的batch size刚好设为1的话就增大一下batch size就好了。

我的batch size的大小是默认的12，但还是会出现那样的问题。

**Comment by cardwing** [comment]
你在测试模型的时候会出现问题吗？还有就是训练模型的时候，你的输入图片的大小是多少？如果你修改了原本的图片大小，那就需要同时修改模型里的全连接层。

**Comment by ygqhfut** [comment]
测试的时候没有问题，无论训练还是测试的图片数据集都是CULane，还有就是如果像1280*720的图片大小，全连接层的尺寸怎么修改呢？

**Comment by cardwing** [comment]
你测试和训练的时候用的图片大小一样吗？你有没有修改训练时的一些参数设置（比如说放缩后的图片大小等）？

**Comment by ygqhfut** [comment]
测试和训练的图片大小都是一样的，也没有修改参数设置，但是总是报错，感觉是torch的问题，但是测试又能跑通。

**Comment by cardwing** [comment]
这个问题我确实没有遇到过。不过看报错的位置，感觉是模型载入的时候出了问题，不然用getParameters()不应该报错。不过，你的测试又能跑通。所以有可能是torch的问题。不过，你只用了一块GPU，而batch size又是12，应该训练不了吧（会超显存）。所以，你将batch size设为8跑一下试试。

**Comment by ygqhfut** [comment]
不行，还是报同样的错误。同样的测试还是没有问题，我把torch重装一下吧

**Comment by ceasorcai** [comment]
@ygqhfut 请问后来解决了吗？我也是同样的问题，谢谢

**Comment by ygqhfut** [comment]
@ceasorcai 没有解决，当我更换了一台服务器发现可以跑通了

**Comment by ceasorcai** [comment]
@ygqhfut 好吧，谢谢啊

**Comment by ygqhfut** [comment]
@cardwing 作者你好，之前因为没有完整的训练好一个epoch，这次我用了两个GPU，训练的脚本如下：
exp=vgg_SCNN_DULR_w9
data=./data/CULane
CUDA_VISIBLE_DEVICES="0,1,2,3" th main.lua \
   -data  ${data} \
   -train ${data}/list/train_gt.txt \
   -val ${data}/list/val_gt.txt \
   -dataset lane \
   -save experiments/models/${exp} \
   -retrain experiments/models/${exp}/ENet_concat_new.t7 \
   -shareGradInput true \
   -nThreads 4 \
   -nGPU 2 \
   -batchSize 6 \
   -maxIter 100000 \
   -LR 0.01 \
   -backWeight 0.4 \
   -nEpochs 100 \
2>&1|tee experiments/models/${exp}/train.log
然后一个epoch训练结束，开始下一个epoch的时候，出现了这样的错误，想请问下原因和解决办法，谢谢
/home/wsco/torch/install/bin/luajit: ...o/torch/install/share/lua/5.1/cunn/DataParallelTable.lua:617: bad argument #1 to 'copy' (sizes do not match at /home/wsco/torch/extra/cutorch/lib/THC/THCTensorCopy.cu:31)
stack traceback:
	[C]: in function 'copy'
	...o/torch/install/share/lua/5.1/cunn/DataParallelTable.lua:617: in function '_concatTensorRecursive'
	...o/torch/install/share/lua/5.1/cunn/DataParallelTable.lua:597: in function '_concatTensorRecursive'
	...o/torch/install/share/lua/5.1/cunn/DataParallelTable.lua:584: in function '_concat'
	...o/torch/install/share/lua/5.1/cunn/DataParallelTable.lua:202: in function 'forward'
	./train.lua:125: in function 'test'
	main.lua:55: in main chunk
	[C]: in function 'dofile'
	...wsco/torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:150: in main chunk
	[C]: at 0x55c82b31d6f0

**Comment by cardwing** [comment]
应该是训练的样本数目不能被batch size整除导致的。你可以在image list里面多重复几张图就好了。

**Comment by ygqhfut** [comment]
的确可以了，谢谢

**Comment by ygqhfut** [comment]
作者你好，当我使用自己的数据集进行训练的时候，一个epoch结束后的Test出错，错误如下：
/home/wsco/torch/install/bin/luajit: /home/wsco/torch/install/share/lua/5.1/threads/threads.lua:183: [thread 4 callback] ./datasets/lane.lua:22: attempt to index local 'label' (a number value)
stack traceback:
	./datasets/lane.lua:22: in function 'get'
	./dataloader.lua:96: in function <./dataloader.lua:92>
	[C]: in function 'xpcall'
	/home/wsco/torch/install/share/lua/5.1/threads/threads.lua:234: in function 'callback'
	/home/wsco/torch/install/share/lua/5.1/threads/queue.lua:65: in function </home/wsco/torch/install/share/lua/5.1/threads/queue.lua:41>
	[C]: in function 'pcall'
	/home/wsco/torch/install/share/lua/5.1/threads/queue.lua:40: in function 'dojob'
	[string "  local Queue = require 'threads.queue'..."]:15: in main chunk
stack traceback:
	[C]: in function 'error'
	/home/wsco/torch/install/share/lua/5.1/threads/threads.lua:183: in function 'dojob'
	./dataloader.lua:157: in function '(for generator)'
	./train.lua:118: in function 'test'
	main.lua:55: in main chunk
	[C]: in function 'dofile'
	...wsco/torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:150: in main chunk
	[C]: at 0x55fbbf5a96f0
load image failed!
load image failed!
load image failed!
这个是什么原因呢？其中我的数据集的分辨率和CULane一样

**Comment by cardwing** [comment]
根据报错信息，你的问题应该是出现在标签读取以后对标签操作的时候。你可以仔细检查一下标签的路径或者是格式。

**Comment by ygqhfut** [comment]
谢谢，我重新更改了标签格式的确可以了

**Comment by redasu** [comment]
sorry for commenting closed ticket, but can author tell what was solution to the original issue? i believe it was reinstalling os, right? what was os, cuda, cudnn etc. that fixed the issue?
